Nomad: recommendations for singleton deployments (#1473)
hashicorp/web-unified-docs: content/nomad/v1.11.x/content/docs/job-declare/strategy/singleton.mdx

workload needs exclusive access to a remote resource like a data store. Nomad
does not support singleton deployments as a built-in feature. Your workloads
continue to run even when the Nomad client agent has crashed, so ensuring
there's at most one allocation for a given workload some cooperation from the
job. This document describes how to implement singleton deployments.

## Design Goals

The configuration described here meets two primary design goals:

* The design will prevent a specific process with a task from running if there
  is another instance of that task running anywhere else on the Nomad cluster.
* Nomad should be able to recover from failure of the task or the node on which
  the task is running with minimal downtime, where "recovery" means that the
  original task should be stopped and that Nomad should schedule a replacement
  task.
* Nomad should minimize false positive detection of failures to avoid
  unnecessary downtime during the cutover.

There's a tradeoff between between recovery speed and false positives. The
faster you make Nomad attempt to recover from failure, the more likely that a

> transient failure causes a replacement to be scheduled and a subsequent

Reviewer: ```suggestion
faster you make Nomad attempt to recover from failure, the more likely that a
transient failure causes Nomad to schedule a replacement and a subsequent
```
active voice nit

downtime.

Note that it's not possible to design a perfectly zero-downtime singleton
allocation in a distributed system. This design will err on the side of
correctness: having 0 or 1 allocations running rather than the incorrect 1 or 2

> allocations running.

Reviewer: ```suggestion
allocation in a distributed system. This design errs on the side of
correctness: having zero or one allocations running rather than the incorrect one or two
allocations running.
```
nit [spell out numbers zero thru nine](https://github.com/hashicorp/web-unified-docs/blob/main/docs/style-guide/numbers-dates-time/words-as-numbers.md)

give up, and the `stop_on_client_after` duration) are the values that can be
tuned to reduce the maximum amount of downtime the application can have.

The Nomad [Locks API][] can support the operations needed. In psuedo-code these
operations are:

* To acquire the lock, `PUT /v1/var/:path?lock-acquire`
  * On success: start heartbeat every 1/2 TTL
  * On conflict or failure: retry with backoff and timeout.
    * Once out of attempts, exit the process with error code.
* To heartbeat, `PUT /v1/var/:path?lock-renew`
  * On success: continue
  * On conflict: exit the process with error code
  * On failure: retry with backoff up to TTL.
    * If TTL expires, attempt to revoke lock, then exit the process with error code.

The allocation can safely use the Nomad [Task API][] socket to write to the
locks API, rather than communicating with the server directly. This reduces load
on the server and speeds up detection of failed client nodes because the
disconnected client cannot forward the Task API requests to the leader.

The [`nomad var lock`][] command implements this logic and can be used to shim
the process being locked.

### ACLs

> Allocations cannot write to Variables by default. You must configure a

Reviewer: ```suggestion
Allocations cannot write to Nomad variables by default. You must configure a
```

Clarification

            "-vv", "-f", "-p", "8001", "-h", "/local" # application args
        ]
      }

      identity {
        env = true
      }
    }
  }
}
```

If you don't want to ship a Nomad binary in the container image you can make a
read-only mount of the binary from a host volume. This will only work in cases
where the Nomad binary has been statically linked or you have glibc in the
container image.

```hcl
job "example" {
  group "group" {

    disconnect {
      stop_on_client_after = "1m"
    }

    volume "binaries" {
      type      = "host"
      source    = "binaries"
      read_only = true
    }

    task "primary" {
      config {
        driver  = "docker"
        image   = "example/app:1"
        command = "/opt/bin/nomad"
        args = [
            "var", "lock", "nomad/jobs/example/lock", # lock
            "busybox", "httpd",                       # application
            "-vv", "-f", "-p", "8001", "-h", "/local" # application args
        ]
      }

      identity {
        env = true # make NOMAD_TOKEN available to lock command
      }

      volume_mount {
        volume      = "binaries"
        destination = "/opt/bin"
      }
    }
  }
}

> ```

Reviewer: ````suggestion
<CodeBlockConfig lineNumbers highlight="8-12,30-33">

```hcl
job "example" {
  group "group" {

    disconnect {
      stop_on_client_after = "1m"
    }

    volume "binaries" {
      type      = "host"
      source    = "binaries"
      read_only = true
    }

    task "primary" {
      config {
        driver  = "docker"
        image   = "example/app:1"
        command = "/opt/bin/nomad"
        args = [
            "var", "lock", "nomad/jobs/example/lock", # lock
            "busybox", "httpd",                       # application
            "-vv", "-f", "-p", "8001", "-h", "/local" # application args
        ]
      }

      identity {
        env = true # make NOMAD_TOKEN available to lock command
      }

      volume_mount {
        volume      = "binaries"
        destination = "/opt/bin"
      }
    }
  }
}
```

</CodeBlockConfig>
````

added code highlights

### Sidecar Lock

If cannot implement the lock logic in your application or with a shim such as

> `nomad var lock`, you'rll need to implement it such that the task you are

Reviewer: ```suggestion
`nomad var lock`, you need to implement it such that the task you are
```

locking is running as a sidecar of the locking task, which has

> [`task.leader=true`][] set.

Reviewer: ```suggestion
### Sidecar lock

If you cannot implement the lock logic in your application or with a shim such
as `nomad var lock`, you need to implement it such that the task you are locking
is running as a sidecar of the locking task, which has [`task.leader=true`][]
set.
```
